### PR TITLE
`eth-json-rpc-provider` migration - Integration into `packages/`

### PR DESCRIPTION
**Pull request by MajorLift**
## Explanation

This PR implements the following incremental steps in the process for migrating `eth-json-rpc-provider` into the core monorepo:

***

#### 5. Port tags 
  - See: https://github.com/MetaMask/core/issues/1800
  
<details>  
  <summary>Push ported tags to core repo</summary>
  
- [x] https://github.com/MetaMask/core/releases/tag/@metamask/eth-json-rpc-provider@2.2.0
- [x] https://github.com/MetaMask/core/releases/tag/@metamask/eth-json-rpc-provider@2.1.0
- [x] https://github.com/MetaMask/core/releases/tag/@metamask/eth-json-rpc-provider@2.0.0
- [x] https://github.com/MetaMask/core/releases/tag/@metamask/eth-json-rpc-provider@1.0.1
- [x] https://github.com/MetaMask/core/releases/tag/@metamask/eth-json-rpc-provider@1.0.0
</details>

<details>
  <summary>Verify that the tag diff links in CHANGELOG are working</summary>
  
- [x] **WONTFIX**: https://github.com/MetaMask/core/compare/@metamask/eth-json-rpc-provider@2.2.0...HEAD
- [x] https://github.com/MetaMask/core/compare/@metamask/eth-json-rpc-provider@2.1.0...@metamask/eth-json-rpc-provider@2.2.0
- [x] https://github.com/MetaMask/core/compare/@metamask/eth-json-rpc-provider@2.0.0...@metamask/eth-json-rpc-provider@2.1.0
- [x] https://github.com/MetaMask/core/compare/@metamask/eth-json-rpc-provider@1.0.1...@metamask/eth-json-rpc-provider@2.0.0
- [x] https://github.com/MetaMask/core/compare/@metamask/eth-json-rpc-provider@1.0.0...@metamask/eth-json-rpc-provider@1.0.1
</details>

#### 1. The big leap
- [x] **Move migration target from `migrated-packages/` to `packages/`.**
- [x] Run `yarn install` in the root directory.
- [x] Check that all tests are passing in migration target by running `yarn workspace @metamask/<package-name> test`.

#### 2. Update downstream repos
- [x] Add tsconfig reference paths for migration target in downstream packages and root.
- [x] Bump migration target version in downstream packages and root.

#### 3. Linter fixes
- [x] Apply yarn constraints fixes to migration target package.json file: `yarn constraints --fix` (run twice).
- [x] Identify validator fixes for CHANGELOG using `yarn workspace @metamask/<package-name> changelog:validate` and apply the diffs. 

#### 4. Resolve downstream errors
- [x] https://github.com/MetaMask/core/pull/1653
  - If introducing the migration target breaks any downstream repos:
    - [x] Resolve simple errors
    - [x] Mark and ignore complex errors using `@ts-expect-error TODO:` annotations. 
  - [x] Create a separate issue for resolving the marked errors as soon as the migration is completed.

#### 5. Finalize merge
- [x] Check that all tests are passing in all subpackages of core and CI.
- [x] Merge `packages/<package-name>` directory into core main branch.

***

See https://github.com/MetaMask/core/issues/1551#issuecomment-1745665740 for an outline of the entire process.

## Next Steps

- The next PR(s) will implement the final steps of the migration process (D-1 in the migration checklist). 

## Blocked by
- Dependencies:
  - [x] typescript bump: https://github.com/MetaMask/core/pull/1718
  - [x] `@metamask/utils` bump: https://github.com/MetaMask/core/pull/1639
- Downstream type errors:
  - [x] https://github.com/MetaMask/core/pull/1653
  - [ ] https://github.com/MetaMask/eth-json-rpc-provider/pull/14 (ignored)
  - [ ] https://github.com/MetaMask/utils/pull/140 (ignored)
- Tag porting:
  - [x] https://github.com/MetaMask/core/pull/1802
  - [x] "Unreleased" tag diff link shows entire history of core: https://github.com/MetaMask/core/compare/@metamask/eth-json-rpc-provider@2.2.0...HEAD

## References

- Contributes to #1685
- Contributes to #1551

### `@metamask/eth-json-rpc-provider`

- **ADDED**: Migrated into the core monorepo.

## Checklist

- [x] I've updated the test suite for new or updated code as appropriate
- [x] I've updated documentation (JSDoc, Markdown, etc.) for new or updated code as appropriate
- [x] I've highlighted breaking changes using the "BREAKING" category above as appropriate

### CODE REVIEW

**Comment by MajorLift** [comment]
@SocketSecurity ignore-all

All packages versions are current with main branch.

File: packages/eth-json-rpc-provider/src/index.test.ts [hunk 1]
```diff
@@ -4,6 +4,7 @@ describe('Package exports', () => {
   it('has expected exports', () => {
     expect(Object.keys(allExports)).toMatchInlineSnapshot(`
       Array [
+        "SafeEventEmitterProvider",
```
Comment (mcmire) [inline packages/eth-json-rpc-provider/src/index.test.ts:7]: What's the reason for this?

File: packages/eth-json-rpc-provider/src/index.ts [hunk 2]
```diff
@@ -1,3 +1,3 @@
 export * from './provider-from-engine';
 export * from './provider-from-middleware';
-export type { SafeEventEmitterProvider } from './safe-event-emitter-provider';
```
Comment (mcmire) [inline packages/eth-json-rpc-provider/src/index.ts:3]: This changes what is being exported (the whole class instead of just the type). What's the reason for this change?

File: packages/network-controller/package.json [hunk 3]
```diff
@@ -33,7 +33,7 @@
     "@metamask/controller-utils": "^5.0.2",
     "@metamask/eth-json-rpc-infura": "^9.0.0",
     "@metamask/eth-json-rpc-middleware": "^12.0.0",
-    "@metamask/eth-json-rpc-provider": "^2.1.0",
+    "@metamask/eth-json-rpc-provider": "^2.2.0",
```
Comment (mcmire) [inline packages/network-controller/package.json:36]: Hmm. Given that we are affecting other packages in this PR, should we bump everything to ^2.2.0 first? I think it'd be better to keep this PR just focused to `eth-json-rpc-provider` if we can.

**Review by mcmire — CHANGES_REQUESTED** [review]
Marking this as "request changes" because of the change to `network-controller` (see comment there).

File: packages/network-controller/package.json (hunk 3 above)
Comment (MajorLift) [reply packages/network-controller/package.json:36]: This is necessary because of the reference path shadowing thing we discussed. Once `eth-json-rpc-provider` is moved into `packages/`, downstream packages can no longer use previous versions by pointing to `node_modules/`, because all imports from `@metamask/eth-json-rpc-provider` will always point to the path in `packages/`, and therefore to the latest version. This seems to happen even if the migrated package isn't added to the reference paths in the downstream pkgs' typescript config files.

File: packages/eth-json-rpc-provider/src/index.test.ts (hunk 1 above)
Comment (MajorLift) [reply packages/eth-json-rpc-provider/src/index.test.ts:7]: This one follows from the export statement change for `SafeEventEmitterProvider` https://github.com/MetaMask/core/pull/1738/files#r1357548245. The test is checking what non-types are being exported from `index.ts`.

File: packages/network-controller/package.json (hunk 3 above)
Comment (mcmire) [reply packages/network-controller/package.json:36]: That makes sense, but isn't this happening because `network-controller` is currently using version 2.1.0? If we bumped that first then the latest version of the package being used and current version of the package as it exists after being migrated would be the same.

File: packages/eth-json-rpc-provider/src/index.ts (hunk 2 above)
Comment (MajorLift) [reply packages/eth-json-rpc-provider/src/index.ts:3]: See: first line of `network-controller/tests/fake-provider`.

File: packages/network-controller/tests/fake-provider.ts [hunk 4]
```diff
@@ -1,4 +1,4 @@
-import { SafeEventEmitterProvider } from '@metamask/eth-json-rpc-provider/dist/safe-event-emitter-provider';
+import { SafeEventEmitterProvider } from '@metamask/eth-json-rpc-provider';
```
Comment (MajorLift) [inline packages/network-controller/tests/fake-provider.ts:1]: See line 85:
```ts
export class FakeProvider extends SafeEventEmitterProvider {
```

Importing `SafeEventEmitterProvider` as a type breaks that line, because it's being used as a class.

This is why the previous import statement pointed directly to `node_modules/.../dist/...` instead of using the `SafeEventEmitterProvider` type exported by the package. 

That import statement can no longer be used (it won't point to `node_modules` with `eth-json-rpc-provider` in `packages/`) which is why the export from `eth-json-rpc-provider` needs to be fixed from type to class.

File: packages/network-controller/package.json (hunk 3 above)
Comment (MajorLift) [reply packages/network-controller/package.json:36]: We could bump to eth-json-rpc-provider to 2.2.0 everywhere first in a separate PR, but that PR wouldn't include the changes being made to `network-controller/` here (the last two commits). 

Because fixing the `../dist/..` import statements requires fixing that export statement in `eth-json-rpc-provider`, we'd still end up with changes being made to both packages simultaneously before the migration can be merged.

Comment (MajorLift) [reply packages/network-controller/package.json:36]: I guess we could change the `SafeEventEmitterProvider` export statement in the original repo and release 2.2.1, but that would mean redoing the migration process from the start.

File: packages/network-controller/tests/fake-provider.ts (hunk 4 above)
Comment (mcmire) [reply packages/network-controller/tests/fake-provider.ts:1]: Ah, I see. Thank you for explaining, somehow that was not connecting.

File: packages/eth-json-rpc-provider/src/index.ts (hunk 2 above)
Comment (mcmire) [reply packages/eth-json-rpc-provider/src/index.ts:3]: Okay, no problem. I guess this wouldn't be a breaking change, so it's okay.

**Comment by MajorLift** [comment]
> Marking this as "request changes" because of the change to `network-controller` (see comment there).

Are there still any changes to the PR you'd like to see, or maybe a different approach we could try for the `network-controller` fixes?

**Review by mcmire — APPROVED** [review]
I'm good with this PR!

**Review by cryptodev-2s — APPROVED** [review]
LGTM!
